File: apps/hexend/src/projectIO.ts
import { readFile, writeFile } from "node:fs/promises";
import { dirname } from "node:path";
import {
  parseHexenProject,
  serializeHexenProject,
  type HexenProject,
  type InlineLocationContent,
} from "@hex-enductor/hexen-schema";
import { createObsidianResolver } from "@hex-enductor/content-obsidian";
import type { ResolvedContent } from "@hex-enductor/content-resolver";

// Inline content needs no I/O and no resolver package of its own — it's
// already exactly a ResolvedContent, just with the `type` discriminant
// stripped off.
function resolveInlineContent(content: InlineLocationContent): ResolvedContent {
  return { title: content.title, body: content.body };
}

export interface OpenedProject {
  project: HexenProject;
  warnings: string[];
  /** location.id -> resolved content, for every location whose content resolved cleanly. */
  resolvedContent: Record<string, ResolvedContent>;
  /** location.id -> the error message, for every location whose content failed to resolve (e.g. a moved/renamed vault file). */
  resolveErrors: Record<string, string>;
}

export async function openProject(path: string): Promise<OpenedProject> {
  const yamlText = await readFile(path, "utf-8");
  const { project, warnings } = parseHexenProject(yamlText);

  const resolvedContent: Record<string, ResolvedContent> = {};
  const resolveErrors: Record<string, string> = {};

  // A Location's own content.type picks how it resolves, independent of
  // the project's — an obsidian-backed project can still hold inline
  // locations that don't warrant a vault file of their own.
  const obsidianResolver =
    project.content.type === "obsidian"
      ? createObsidianResolver({ projectDir: dirname(path), vaultRoot: project.content.vaultRoot })
      : null;

  await Promise.all(
    project.locations
      .filter((location) => location.content !== null)
      .map(async (location) => {
        // location.content is narrowed non-null by the filter above
        const content = location.content!;
        try {
          if (content.type === "inline") {
            resolvedContent[location.id] = resolveInlineContent(content);
          } else if (obsidianResolver) {
            resolvedContent[location.id] = await obsidianResolver.resolve(content);
          } else {
            throw new Error(`Location "${location.id}" has obsidian content, but this project has no vault configured`);
          }
        } catch (err) {
          resolveErrors[location.id] = err instanceof Error ? err.message : String(err);
        }
      }),
  );

  return { project, warnings, resolvedContent, resolveErrors };
}

export async function saveProject(path: string, project: HexenProject): Promise<void> {
  const yamlText = serializeHexenProject(project);
  await writeFile(path, yamlText, "utf-8");
}

File: apps/hexend/src/project_io.rs
//! Opening and saving a project on disk. Parsing, serializing and the
//! cross-reference check are crates/project-ops's `document`.

use std::collections::HashMap;
use std::path::Path;

use thiserror::Error;

use crate::pb::hexen::v1::{location_content, project_content, HexenProject, ResolvedContent};
use crate::resolver::{ObsidianResolver, ObsidianResolverConfig};
pub use project_ops::document::{parse_hexen_project, serialize_hexen_project, ParseError, ParsedHexenProject};
use project_ops::content::{no_vault_error, resolve_inline_content};

pub struct OpenedProject {
    pub project: HexenProject,
    pub warnings: Vec<String>,
    /// location.id -> resolved content, for every location whose content resolved cleanly.
    pub resolved_content: HashMap<String, ResolvedContent>,
    /// location.id -> the error message, for every location whose content failed to resolve.
    pub resolve_errors: HashMap<String, String>,
}

#[derive(Debug, Error)]
pub enum OpenError {
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    Parse(#[from] ParseError),
}

pub async fn open_project(path: &Path) -> Result<OpenedProject, OpenError> {
    let yaml_text = tokio::fs::read_to_string(path).await?;
    let ParsedHexenProject { project, warnings } = parse_hexen_project(&yaml_text)?;

    let mut resolved_content = HashMap::new();
    let mut resolve_errors = HashMap::new();

    // A Location's own content.kind picks how it resolves, independent
    // of the project's — an obsidian-backed project can still hold
    // inline locations that don't warrant a vault file of their own.
    let obsidian_resolver = match &project.content {
        Some(pc) => match &pc.kind {
            Some(project_content::Kind::Obsidian(o)) => {
                let project_dir = path.parent().unwrap_or_else(|| Path::new(".")).to_path_buf();
                Some(ObsidianResolver::new(ObsidianResolverConfig {
                    project_dir,
                    vault_root: o.vault_root.clone(),
                }))
            }
            _ => None,
        },
        None => None,
    };

    for location in &project.locations {
        let Some(content) = &location.content else {
            continue;
        };

        let result: Result<ResolvedContent, String> = match &content.kind {
            Some(location_content::Kind::Inline(inline)) => Ok(resolve_inline_content(inline)),
            Some(location_content::Kind::Obsidian(obsidian)) => match &obsidian_resolver {
                Some(resolver) => resolver.resolve(obsidian).await.map_err(|e| e.to_string()),
                None => Err(no_vault_error(&location.id)),
            },
            None => continue,
        };

        match result {
            Ok(resolved) => {
                resolved_content.insert(location.id.clone(), resolved);
            }
            Err(err) => {
                resolve_errors.insert(location.id.clone(), err);
            }
        }
    }

    Ok(OpenedProject {
        project,
        warnings,
        resolved_content,
        resolve_errors,
    })
}

pub async fn save_project(path: &Path, project: &HexenProject) -> Result<(), OpenError> {
    let yaml_text = serialize_hexen_project(project).map_err(ParseError::Yaml)?;
    tokio::fs::write(path, yaml_text).await?;
    Ok(())
}
